In [8]:
import numpy as np
import matplotlib.pyplot as plt
from nbody import Particles, NBodySimulator
from nbody import load_files, save_movie

# from numba import set_num_threads

In [9]:
# Uncomment the following line to install required packages if needed

#!conda install -y -c conda-forge ffmpeg, glob, numba

# N-Body Simulation

In this notebook, we will test the `NBodySimulator` class in `./nbody/simulator.py`. 

In [10]:
# Set the number of threads to use for numba
# nthreads = 1
# set_num_threads(nthreads)

# Exercise 1: The Sun-Earth System

The first test is to implment the Sun-Earth system, which is a two body system. We need to make sure that you could simulate a circular motion first.

## Step 1. The initial condition

Copy the initial condition of the Earth-Sun system from `project2_demo1.ipynb`.

In [11]:
# TODO:

num_particles = 2  # number of particles
particles = Particles(N=num_particles)
masses        = np.ones((num_particles,1))
positions     = np.zeros((num_particles,3)) # 3 directions
velocities    = np.zeros((num_particles,3))
accelerations = np.zeros((num_particles,3))
tags          = np.linspace(1,num_particles,num_particles)







## Step 2. N-body simulation

Make sure you have implmented the below two methods `setup()` and `evolve()`. Run the n-body simulation by:

In [12]:
simulation = NBodySimulator(particles=particles)
simulation.setup(G=6.674e-11,rsoft=1e9,method='RK4', io_freq=30)
simulation.evolve(dt=8640, tmax=365*86400)

n= 0 Time:  0  dt:  8640
n= 30 Time:  259200  dt:  8640
n= 60 Time:  518400  dt:  8640
n= 90 Time:  777600  dt:  8640
n= 120 Time:  1036800  dt:  8640
n= 150 Time:  1296000  dt:  8640
n= 180 Time:  1555200  dt:  8640
n= 210 Time:  1814400  dt:  8640
n= 240 Time:  2073600  dt:  8640
n= 270 Time:  2332800  dt:  8640
n= 300 Time:  2592000  dt:  8640
n= 330 Time:  2851200  dt:  8640
n= 360 Time:  3110400  dt:  8640
n= 390 Time:  3369600  dt:  8640
n= 420 Time:  3628800  dt:  8640
n= 450 Time:  3888000  dt:  8640
n= 480 Time:  4147200  dt:  8640
n= 510 Time:  4406400  dt:  8640
n= 540 Time:  4665600  dt:  8640
n= 570 Time:  4924800  dt:  8640
n= 600 Time:  5184000  dt:  8640
n= 630 Time:  5443200  dt:  8640
n= 660 Time:  5702400  dt:  8640
n= 690 Time:  5961600  dt:  8640
n= 720 Time:  6220800  dt:  8640
n= 750 Time:  6480000  dt:  8640
n= 780 Time:  6739200  dt:  8640
n= 810 Time:  6998400  dt:  8640
n= 840 Time:  7257600  dt:  8640
n= 870 Time:  7516800  dt:  8640
n= 900 Time:  7776000  d

## Step 3. Visualization

Check the code in `./nobdy/visualization.py`.
Data loader is implmented in the function `load_files`.

In [13]:
fns = load_files('nbody')
print(fns)


['data_nbody\\nbody_000000.dat', 'data_nbody\\nbody_000030.dat', 'data_nbody\\nbody_000060.dat', 'data_nbody\\nbody_000090.dat', 'data_nbody\\nbody_000120.dat', 'data_nbody\\nbody_000150.dat', 'data_nbody\\nbody_000180.dat', 'data_nbody\\nbody_000210.dat', 'data_nbody\\nbody_000240.dat', 'data_nbody\\nbody_000270.dat', 'data_nbody\\nbody_000300.dat', 'data_nbody\\nbody_000330.dat', 'data_nbody\\nbody_000360.dat', 'data_nbody\\nbody_000390.dat', 'data_nbody\\nbody_000420.dat', 'data_nbody\\nbody_000450.dat', 'data_nbody\\nbody_000480.dat', 'data_nbody\\nbody_000510.dat', 'data_nbody\\nbody_000540.dat', 'data_nbody\\nbody_000570.dat', 'data_nbody\\nbody_000600.dat', 'data_nbody\\nbody_000630.dat', 'data_nbody\\nbody_000660.dat', 'data_nbody\\nbody_000690.dat', 'data_nbody\\nbody_000720.dat', 'data_nbody\\nbody_000750.dat', 'data_nbody\\nbody_000780.dat', 'data_nbody\\nbody_000810.dat', 'data_nbody\\nbody_000840.dat', 'data_nbody\\nbody_000870.dat', 'data_nbody\\nbody_000900.dat', 'data_n

Modify the matplotlib script 

In [15]:
import ffmpeg

au = 1.496e11
save_movie(fns, lengthscale=2*au,filename='nbody_earth_sun.mp4', fps=10)

ModuleNotFoundError: No module named 'ffmpeg'

# Exercise 2: N-Body simulation

Now, let's test our n-body solver with more particles. Copy the initial condition from the Exercise 2 in `project2_demo1.ipynb`.

In [ ]:
# TODO










## Step 2. Simulation.

Visualize the system but use dimensionless units. Set `G=1` and `rsoft=0.001`.

In [ ]:
simulation = NBodySimulator(particles=pts)
simulation.setup(G=1,rsoft=0.001,method='RK4', io_freq=10, io_header='cluster')
simulation.evolve(dt=0.01, tmax=1)

It is VERY slow !!!!

Beacuse the accerlation calculations requires order of NxN calculations!
Could we improve it?

First, we should avoid the two "for" loops in the calculation using python.\
Second, we have many cup cores in our machine. Could we parallelize it?

# `numba`

Step 1. Take out the function to calculate the accerlation from the class into a pure function that only counts math calculations (aka. a kernel).

Step 2. Use the `numba`'s `@jit` micro. 

Results: In my experiments. I got roughly x140 speed up with the @jit(nopython=True). 

Step 3. Try using `@njit` and `prange` in numba.

## Performance

### Strong Scaling Test

**Strong scaling**: The number of processors is increased while the problem size remains constant.\
We could use `%timeit` to measure the performance. Let's run 1000 particles for 10 step.

In [ ]:
num_particles = 1000
pts = Particles(N=num_particles)
pts.masses = np.ones((num_particles, 1))* 10/num_particles
pts.positions = np.random.randn(num_particles, 3)
pts.velocities = np.random.randn(num_particles, 3)
pts.accelerations = np.zeros((num_particles, 3))

simulation = NBodySimulator(particles=pts)
simulation.setup(G=1,rsoft=0.001,method='RK4', io_screen=False, io_freq=0, io_header='cluster')
#simulation.evolve(dt=0.02, tmax=1)

In [ ]:
%timeit simulation.evolve(dt=0.02, tmax=0.2)

**Kuo-Chuan's measurements**

CPU: 3 GHz 10-Core Intel Xeon W

1000 particles. 10 steps.

without `numba`: 5min 31s ± 2 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
with `numba`:
* 1 core (`jit`): 17.7 s ± 328 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* 1 core (`njit`): 10.4 s ± 337 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* 2 cores: 7.61 s ± 52.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* 6 cores: 3.35 s ± 50.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* 10 cores: 2.12 s ± 16.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* 12 cores: 1.95 s ± 36.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* 16 cores: 1.64 s ± 46 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* 20 cores: 1.49 s ± 55.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [ ]:
threads = [1,2,6,10,12,16,20]                     # Replace it to your measured values
times = [10.4, 7.61, 3.35, 2.12, 1.95, 1.64,1.49] # Replace it to your measured values
ideal = times[0]/np.array(threads)
plt.loglog(threads, times)
plt.loglog(threads, ideal, '--')
plt.xlabel('Number of threads')
plt.ylabel('Time for 10 iterations (s)')
plt.legend(['Measured', 'Ideal'])
plt.title('Strong scaling')
plt.show()

## Weak Scaling Test

In N-body simulation, the problem size is proportional to $N^2$.\
**Weaking scaling** test measures the scaling with the same problem size per thread (core).

In [ ]:
num_particles = int(225*np.sqrt(nthreads))
print("N =", num_particles, ", threads = ", nthreads)
pts = Particles(N=num_particles)
pts.masses = np.ones((num_particles, 1))* 10/num_particles
pts.positions = np.random.randn(num_particles, 3)
pts.velocities = np.random.randn(num_particles, 3)
pts.accelerations = np.zeros((num_particles, 3))

simulation = NBodySimulator(particles=pts)
simulation.setup(G=1,rsoft=0.001,method='RK4', io_screen=False, io_freq=0, io_header='cluster')

In [ ]:
%timeit simulation.evolve(dt=0.02, tmax=0.2)

Kuo-Chuan's measurements

* N = 225 threads =  1: 537 ms ± 13.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* N = 450 threads =  4: 981 ms ± 17.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* N = 636 threads =  8: 1.07 s ± 22.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* N = 779 threads =  12: 1.17 s ± 33.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* N = 900 threads =  16: 1.29 s ± 6.85 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
* N = 1006 threads =  20: 1.51 s ± 35.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [ ]:
threads = [1,4,8,12,16,20]                     # Replace it to your measured values
times = [0.537, 0.981, 1.07, 1.17, 1.29, 1.51] # Replace it to your measured values
ideal = 0.981 * np.ones(len(threads))
plt.plot(threads, times)
plt.plot(threads, ideal, '--')
plt.xlabel('Number of threads')
plt.ylabel('Time for 10 iterations (s)')
plt.legend(['Measured', 'Ideal'])
plt.title('Weak scaling')
plt.show()